metrics-tracing-context/tests/integration.rs

```diff
@@ -110,7 +117,137 @@ fn test_basic_functionality() {
             None,
             DebugValue::Counter(1),
         )]
-    )
+    );
+}
+
+#[test]
+fn test_basic_functionality_record() {
+    let (_guard, snapshotter) = setup(TracingContextLayer::all());
+
+    let user = "ferris";
+    let email = "ferris@rust-lang.org";
+    let span = span!(
+        Level::TRACE,
+        "login",
+        user,
+        user.email = email,
+        user.id = tracing_core::field::Empty,
+    );
+    let _guard = span.enter();
+
+    span.record("user.id", 42);
+    counter!("login_attempts", "service" => "login_service").increment(1);
+
+    let snapshot = snapshotter.snapshot().into_vec();
+
+    assert_eq!(
+        snapshot,
+        vec![(
+            CompositeKey::new(
+                MetricKind::Counter,
+                Key::from_static_parts(LOGIN_ATTEMPTS, SVC_USER_EMAIL_ID)
+            ),
+            None,
+            None,
+            DebugValue::Counter(1),
+        )]
+    );
+}
+
+#[test]
+fn test_basic_functionality_then_record() {
+    let (_guard, snapshotter) = setup(TracingContextLayer::all());
+
+    let user = "ferris";
+    let email = "ferris@rust-lang.org";
+    let span = span!(
+        Level::TRACE,
+        "login",
+        user,
+        user.email = email,
+        user.id = tracing_core::field::Empty,
+    );
+    let _guard = span.enter();
+    let mut snapshots = vec![];
+    {
+        counter!("login_attempts", "service" => "login_service").increment(1);
+
+        let snapshot = snapshotter.snapshot().into_vec();
+
+        snapshots.push((
+            CompositeKey::new(
+                MetricKind::Counter,
+                Key::from_static_parts(LOGIN_ATTEMPTS, SVC_USER_EMAIL),
+            ),
+            None,
+            None,
+            DebugValue::Counter(1),
+        ));
+
+        assert_eq!(snapshot, snapshots);
+    }
+    span.record("user.id", 42);
+    {
+        counter!("login_attempts", "service" => "login_service").increment(1);
+
+        let snapshot = snapshotter.snapshot().into_vec();
+
+        snapshots.push((
+            CompositeKey::new(
+                MetricKind::Counter,
+                Key::from_static_parts(LOGIN_ATTEMPTS, SVC_USER_EMAIL_ID),
+            ),
+            None,
+            None,
+            DebugValue::Counter(1),
+        ));
+
+        assert_eq!(snapshot, snapshots);
+    }
+}
+
+#[test]
+fn test_rerecord() {
+    static USER_ID_42: &[Label] = &[Label::from_static_parts("user.id", "42")];
+    static USER_ID_123: &[Label] =
+        &[Label::from_static_parts("user.id", "42"), Label::from_static_parts("user.id", "123")];
```

Comment:
If the field is `record`ed more than one time, each time emitting a metric, we would record every seen value. Is that a concern? Probably should be documented somewhere?

Reply:
`tracing`, in one (or many) of their `Subscriber` `Layer`s, currently has the same problem, here's the open issue: https://github.com/tokio-rs/tracing/issues/1372#issuecomment-1709274609.

Reply:
One obvious downside is something like this:
```rust
let span = tracing::info_span!("loop", item = field::Empty);
let _g = span.enter();
for i in foo() {
	span.record("item", i);
	metrics::counter!("counter").increment(1);
}
```
Would cause us to extend `Labels` with each iteration.

Reply:
We could store `Labels` as `pub struct Labels(pub LinearOwnedReusable<IndexMap<SharedString, Label>>);`. That would be a breaking change for two reasons: 1) `Labels` are `pub` (despite the `doc(hidden)`); 2) Nested spans  with same field name would overwrite each-other, as opposed to current behavior (`test_nested_spans` test).

Reply:
Yeah, this is all a little bit tricky.

I don't love the idea of having to also store a map for every span, but there's obviously got to be some deduplication logic _somewhere_. I'm just trying to think through if it should live in the layer logic (so that we amortize it across all `record` calls, etc) or if it should happen when we want to collect/build the final set of labels before creating the new metric key. 🤔

Reply:
Let's evaluate the trade-off if storing the map is undesirable:

1. Deduplicate in `on_record`:
    `+`: we store only required labels, saving on allocations.
	`-`: possibly a lot of extra operations on a "happy path" when user does not re-record the fields.
	`+`: key creation becomes free
2. Deduplicate on key creation:
	`+`: it would be the only time we need to deduplicate labels.
	`-`: we store all previous `record`ed values for a label, extending the underlying `Vec` each time.

Reply:
Thinking more about it, your point about extending the labels on each iteration of the loop in your example got me to thinking: even if we were willing to eat the cost of duplicate entries... the more risky part is that we could end up extending that array of labels in an infinite/unbounded fashion. For example, if your loop was an asynchronous iterator, and recorded a field every single time, that stream may never end, etc.

I'd say let's go with the `IndexMap<K, V>`-based approach.